src/main/scala/coreplex/MemoryBus.scala

```diff
@@ -0,0 +1,72 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.coreplex
+
+import Chisel._
+import freechips.rocketchip.config._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util._
+
+// TODO: applies to all caches, for now
+case object CacheBlockBytes extends Field[Int]
+
+/** L2 Broadcast Hub configuration */
+case class BroadcastParams(
+  nTrackers:  Int     = 4,
+  bufferless: Boolean = false)
+case object BroadcastParams extends Field[BroadcastParams]
+
+/** L2 memory subsystem configuration */
+case class BankedL2Params(
+  nMemoryChannels:  Int = 1,
+  nBanksPerChannel: Int = 1,
+  coherenceManager: (Parameters, HasMemoryBus) => (TLInwardNode, TLOutwardNode) = { case (q, _) =>
+    implicit val p = q
+    val MemoryBusParams(_, blockBytes, _, _, _) = p(MemoryBusParams)
+    val BroadcastParams(nTrackers, bufferless) = p(BroadcastParams)
+    val bh = LazyModule(new TLBroadcast(blockBytes, nTrackers, bufferless))
+    (bh.node, bh.node)
+  }) {
+  val nBanks = nMemoryChannels*nBanksPerChannel
+}
+case object BankedL2Params extends Field[BankedL2Params]
+
+/** Parameterization of the memory-side bus created for each memory channel */
+case class MemoryBusParams(
+  beatBytes: Int,
+  blockBytes: Int,
+  masterBuffering: BufferParams = BufferParams.none,
+  slaveBuffering: BufferParams = BufferParams.none,
+  masterFIFOPolicy: TLFIFOFixer.Policy = TLFIFOFixer.all
+) extends TLBusParams
+
+case object MemoryBusParams extends Field[MemoryBusParams]
+
+/** Wrapper for creating TL nodes from a bus connected to the back of each mem channel */
+class MemoryBus(params: MemoryBusParams)(implicit p: Parameters) extends TLBusWrapper(params)(p)
+
+trait HasMemoryBus extends HasSystemBus {
+  private val mbusParams = p(MemoryBusParams)
+  private val MemoryBusParams(beatBytes, blockBytes, _, _, _) = mbusParams
+  private val l2Params = p(BankedL2Params)
+  val BankedL2Params(nMemoryChannels, nBanksPerChannel, coherenceManager) = l2Params
+  val nBanks = l2Params.nBanks
```

Comment:
`nBanks` is a derived parameter, so you can't name it via the case class extractor pattern on the previous line.